### ISSUE TITLE
Debugging Callback queries seems impossible

### ISSUE THREAD

**Issue by Skowt** [telegram-bot-rb/telegram-bot#83]
I've been adding specs for my tests and I currently have a list of callbacks that look like this:

![image](https://user-images.githubusercontent.com/4960236/40272600-51e6dac4-5bb0-11e8-829a-8015336d7d99.png)

I'm trying to write specs to make sure that when different callbacks are received, they trigger the appropriate method. I'm trying to do that like this:

```
describe 'for unsupported callback query, :callback_query' do
    let(:data) { 'afasufsaofi' }
    it 'returns an error' do 
      expect { subject}.to respond_with_message t('.not_supported')
    end
  end
```

But I keep getting:
`expected to make exactly 1 #<Telegram::Bot::ClientStub#70132747235740()>.sendMessage requests, with ..., but made 0, and 0 with`

I can't find anyway of debugging the `subject` as it just returns a `200`. I've tried `.inspect` `.call` and putting a `binding.pry` in the controller but nothing seems to work. Am I going about testing this the wrong way?

**Comment by printercu** [comment]
Hi! `:callback_query` in `describe` should be next arg, not a part of description: https://github.com/telegram-bot-rb/telegram_bot_app/blob/master/spec/requests/telegram_webhooks_spec.rb#L67 . This way it's treated as tag, and rspec includes shared context for it: https://github.com/telegram-bot-rb/telegram-bot/blob/master/lib/telegram/bot/rspec/integration.rb#L47

Make sure you understand that `respond_with_message` matcher tests for exact `respond_with` call: https://github.com/telegram-bot-rb/telegram-bot/blob/master/lib/telegram/bot/rspec/integration.rb#L41. While you use `answer_callback_query` you should use corresponding matcher. This matchers work more like rspec-mocks: first you run code, payload from all requests is saved, and then you test "ensure that bot have made this request with this args": https://github.com/telegram-bot-rb/telegram-bot/blob/master/lib/telegram/bot/rspec/client_matchers.rb. Make sure you have configured bot mocks first to prevent them from real requests: https://github.com/telegram-bot-rb/telegram-bot#testing

PS. There is `CallbackQueryContext` to avoid such long `case`: https://github.com/telegram-bot-rb/telegram-bot#callback-queries. When you don't want to pass any args, just append colon: `data: "settings:"`

**Comment by Skowt** [comment]
**>Hi! :callback_query in describe should be next arg, not a part of description:** 
Yeah sorry I fixed this after posting this issue 🤦‍♂️ 

**>Make sure you understand that respond_with_message matcher tests for exact respond_with call:** 
Makes sense 👌 The code I had in place was just a placeholder so I could try find a way of debugging the `subject` and see what it was doing. 

**>first you run code, payload from all requests is saved,**
How do I know which requests are saved and how would this work on my staging or production servers? Is there a way of choosing which requests are saved and how would you go about setting specific params if it essentially just returns the response based on when you ran the code? Is it all based on the 'data' attribute that's passed? Sorry if I'm not understanding you fully.

* * * 

**>PS. There is CallbackQueryContext to avoid such long case: https://github.com/telegram-bot-rb/telegram-bot#callback-queries. When you don't want to pass any args, just append colon: data: "settings:"**
Thanks so much for this. I got confused with this code and [context_to_action!](https://github.com/telegram-bot-rb/telegram-bot#message-context) feature for trying to split out commands to methods. I'm guessing these could both exist but then you'd end up having duplicate methods. Say I could do the same function with a callback query or a command, would I then just use:
```
context_handler :my_action do |*words|
    my_action_callback_query words[0]
    respond_with :message, text: 'Done!'
  end
```
To just need the callback query method and not a separate one to handle the command?

**Comment by Skowt** [comment]
Just an update, so I was able `binding.pry` working with the request specs. I had one set in the request spec itself and in the controller and I think that prevented the controller pry from working. I removed the one in the spec and then I was able to dig through it. 

To me it just looks like the default calls are being used that are setup in https://github.com/telegram-bot-rb/telegram-bot/blob/6a5e642470d41214719d05aa8889c2f57f913627/lib/telegram/bot/rspec/integration.rb#L2-L19 

I don't see how the calls are recorded the first time they're used but I might be reading that wrong. 

I also don't see how you'd setup a controller spec for this or what it's purpose would be 🤔If I try the usual `get :mycallback` I get a `undefined method get` error and if I try use `process_update` like in the example on the README, I also just get an unknown error.

**Comment by printercu** [comment]
> How do I know which requests are saved and how would this work on my staging or production servers?

Please check out "testing" section in readme https://github.com/telegram-bot-rb/telegram-bot#testing Let me know if something remains unclear.

**Comment by printercu** [comment]
> I got confused with this code and context_to_action!

MessageContext is for `message` updates, and CallbackQueryContext is for `callback_query`. They don't related to each other. MessageContext supports multiple ways - either pass the block for handler, or use other defined method. You can use any existing methods. Think of it the same way as action callbacks:

```
before_action :some_method
context_handler :some_method # invokes some_method if context is some_method

before_action { your_code_goes_here }
context_handler(:some_method) { code }
```

**Comment by printercu** [comment]
It's better proceed with integration tests. It works out of the box in rails app with webhooks mode. If you use only poller mode, you can add bot's route with `if Rails.env.test?` to use integration specs. Without rails it's a little tricky to get integration tests working, so it's easier to go with controller tests. I'll try to find the better way how improve and unify it.

**Comment by printercu** [comment]
I've already slightly updated readme, please see https://github.com/telegram-bot-rb/telegram-bot/tree/unify_specs#testing

**Comment by printercu** [comment]
Did this help?

**Comment by Skowt** [comment]
Hey Sorry for the silence! Going to dig into this today or tomorrow and I'll give a full update 👌 Can close for now if you're trying to clear out issues.

**Comment by printercu** [comment]
It's ok. let's close once it's solved.

**Comment by printercu** [comment]
I'm suggesting using https://github.com/telegram-bot-rb/telegram-bot/tree/unify_specs branch, i'm going to merge it soon.
It has some minor breaking changes (see changelog), but provides consistent interface for testing bots in different applications.

**Comment by Skowt** [comment]
Awesome 🙌 I'm going to do some more testing with my request specs tonight and see how far I get. Hopefully I'm able to make some more progress 💪

**Comment by Skowt** [comment]
Left a comment on your branch 🙂

**Comment by Skowt** [comment]
Closing since everything is working as expected 👌 Ended up using request specs and then throwing a `binding.pry` on the controller at parts where I needed to test. Smooth sailing so far!

**Comment by printercu** [comment]
I've recently released 0.14.0 with changes from #84 #88 #85. I appreciate any feedback if you want to give it a try.

**Comment by Skowt** [comment]
Definitely going to give it a go when I have a chance 🙂 👌
